src/boot/pe.c

```diff
@@ -441,26 +468,89 @@ EFI_STATUS pe_kernel_info(const void *base, uint32_t *ret_compat_address, size_t
 
+        uint64_t image_base;
+        switch (pe->OptionalHeader.Magic) {
+        case OPTHDR32_MAGIC:
+                image_base = pe->OptionalHeader.ImageBase32;
+                break;
+        case OPTHDR64_MAGIC:
+                image_base = pe->OptionalHeader.ImageBase64;
+                break;
+        default:
+                assert_not_reached();
+        }
+
         /* When allocating we need to also consider the virtual/uninitialized data sections, so parse it out
          * of the SizeOfImage field in the PE header and return it */
-        if (ret_size_in_memory)
-                *ret_size_in_memory = pe->OptionalHeader.SizeOfImage;
+        size_t size_in_memory = pe->OptionalHeader.SizeOfImage;
 
         /* Support for LINUX_INITRD_MEDIA_GUID was added in kernel stub 1.0. */
         if (pe->OptionalHeader.MajorImageVersion < 1)
                 return EFI_UNSUPPORTED;
 
         if (pe->FileHeader.Machine == TARGET_MACHINE_TYPE) {
-                *ret_compat_address = 0;
+                if (ret_entry_point)
+                        *ret_entry_point = pe->OptionalHeader.AddressOfEntryPoint;
+                if (ret_compat_entry_point)
+                        *ret_compat_entry_point = 0;
+                if (ret_image_base)
+                        *ret_image_base = image_base;
+                if (ret_size_in_memory)
+                        *ret_size_in_memory = size_in_memory;
                 return EFI_SUCCESS;
         }
 
-        uint32_t compat_address = get_compatibility_entry_address(dos, pe);
-        if (compat_address == 0)
+        uint32_t compat_entry_point = get_compatibility_entry_address(dos, pe);
+        if (compat_entry_point == 0)
                 /* Image type not supported and no compat entry found. */
                 return EFI_UNSUPPORTED;
 
-        *ret_compat_address = compat_address;
+        if (ret_entry_point)
+                *ret_entry_point = 0;
+        if (ret_compat_entry_point)
+                *ret_compat_entry_point = compat_entry_point;
+        if (ret_image_base)
+                *ret_image_base = image_base;
+        if (ret_size_in_memory)
+                *ret_size_in_memory = size_in_memory;
+
+        return EFI_SUCCESS;
+}
+
+/* https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#optional-header-data-directories-image-only */
+#define BASE_RELOCATION_TABLE_DATA_DIRECTORY_ENTRY 5
+
+/* We do not expect PE inner kernels to have any relocations. However that might be wrong for some
+ * architectures, or it might change in the future. If the case of relocation arise, we should transform this
+ * function in a function applying the relocations. However for now, since it would not be exercised and
+ * would bitrot, we leave it as a check that relocations are never expected.
+ */
+EFI_STATUS pe_kernel_check_no_relocation(const void *base) {
+        assert(base);
+
+        const DosFileHeader *dos = (const DosFileHeader *) base;
```

Comment:
the cast is redundant. const void* casts implicitly to any other const pointer type